### PR TITLE
FIX: Preload fonts before rerendering wizard style canvas

### PR DESCRIPTION
**Pull request by martin-brennan**
When changing fonts in the `/wizard/steps/styling` step of
the wizard, users would not see the font loaded straight away,
having to switch to another one then back to the original to
see the result. This is because we are using canvas to render
the style preview and this fails with a Chrome-based intervention
when font loading is taking too long:

> [Intervention] Slow network is detected. See
https://www.chromestatus.com/feature/5636954674692096 for more details.
Fallback font will be used while loading:
https://sea2.discourse-cdn.com/business7/fonts/Roboto-Bold.ttf?v=0.0.9

We can get around this by manually loading the fonts selected using
the FontFace JS API when the user selects them and before rerendering
the canvas. This just requires preloading more information about the
fonts if the user is admin so the wizard can query this data.

image

https://github.com/discourse/discourse/assets/920448/11113752-3764-416f-bd48-0f9ff294a87b

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1065,4 +1068,20 @@ def link_preload
...
+        if font[:variants]
+          font_map_hash[font[:key]] = font[:variants].map do |v|
+            {
+              url: "#{Discourse.base_url}/fonts/#{v[:filename]}?v=#{DiscourseFonts::VERSION}",
+              weight: v[:weight],
+            }
+          end
+        end
+        font_map_hash
+      end
```
Comment (CvX) [inline app/controllers/application_controller.rb:1085]: Warning, I'm codegolfing here 😄 

```suggestion
      .each_with_object({}) do |font, font_map|
        next if !font[:variants]
        font_map[font[:key]] = font[:variants].map do |v|
          {
            url: "#{Discourse.base_url}/fonts/#{v[:filename]}?v=#{DiscourseFonts::VERSION}",
            weight: v[:weight],
          }
        end
      end
```

File: app/assets/javascripts/wizard/addon/components/wizard-preview-base.js [hunk 2]
```diff
@@ -54,21 +55,74 @@ export default Component.extend({
...
 
-  @observes(
-    "step.fieldsById.{color_scheme,body_font,heading_font,homepage_style}.value"
-  )
+  @observes("step.fieldsById.{color_scheme,homepage_style}.value")
   themeChanged() {
     this.triggerRepaint();
   },
 
+  @observes("step.fieldsById.{body_font}.value")
```
Comment (CvX) [inline app/assets/javascripts/wizard/addon/components/wizard-preview-base.js:71]: I'm accepting new observers here for now 😛 we'll _glimmy_ that component at some point 😛

File: app/assets/javascripts/wizard/addon/components/wizard-preview-base.js [hunk 3]
```diff
@@ -54,21 +55,74 @@ export default Component.extend({
...
+  },
+
+  loadFontVariants(font) {
+    const fontVariantData = this.fontMap[font.id];
+
+    // System font for example does not need to load from a remote source.
+    if (fontVariantData && !this.loadedFonts.has(font.id)) {
+      this.loadingFontVariants = true;
+      const fontFaces = fontVariantData.map((fontVariant) => {
+        return new FontFace(`${font.label}`, `url(${fontVariant.url})`, {
```
Comment (CvX) [inline app/assets/javascripts/wizard/addon/components/wizard-preview-base.js:92]:
```suggestion
        return new FontFace(font.label, `url(${fontVariant.url})`, {
```

**Review by CvX — APPROVED** [review]

File: app/assets/javascripts/wizard/addon/components/wizard-preview-base.js (hunk 2 above)
Comment (martin-brennan) [reply app/assets/javascripts/wizard/addon/components/wizard-preview-base.js:71]: Yeah I didn't like it either but didn't want to refactor the whole thing heh

File: app/assets/javascripts/wizard/addon/components/wizard-preview-base.js (hunk 3 above)
Comment (martin-brennan) [reply app/assets/javascripts/wizard/addon/components/wizard-preview-base.js:92]: Ah good catch I was concat'ing something there previously, will remove

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (martin-brennan) [reply app/controllers/application_controller.rb:1085]: Nothing wrong with a _little_ golf, just some putt-putt 😆

**Pull request merged**